File: BOJ/1912.cpp
#include<cstdio>
#include<algorithm>
using namespace std;

int arr[2][100000];

int main() {
	int n;
	scanf("%d", &n);
	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[0][i]);
	}

	arr[1][0] = arr[0][0];
	int MAX = arr[0][0];
	for (int i = 1; i < n; i++) {
		arr[1][i] = max(arr[1][i - 1] + arr[0][i], arr[0][i]);
		MAX = max(MAX, arr[1][i]);
	}

	printf("%d\n", MAX);

	return 0;
}

File: BOJ/python/1912.py
"""
dp[i]: i번째 위치에서 구할 수 있는 가장 큰 값.
(이전에 값들을 모두 더한 값, 현재 값)을 비교하여 더 큰 값으로 업데이트 한다.
"""
import sys

input = sys.stdin.readline


def main() -> None:
    n = int(input().rstrip('\n'))
    arr = list(map(int, input().rstrip('\n').split(' ')))
    dp = [0] * n
    dp[0] = arr[0]

    for i in range(1, n):
        dp[i] = max(dp[i - 1] + arr[i], arr[i])  # dp[i-1]이 음수인 경우 더할 필요가 없다

    # print(max(dp))
    print(dp)


if __name__ == "__main__":
    main()
